### PR TITLE
Repository and ViewModel unit tests

### PR DESCRIPTION
**Pull request by andersonluisp**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -75,6 +75,12 @@ dependencies {
     implementation "org.jetbrains.kotlinx:kotlinx-coroutines-android:$coroutineVersion"
```
Comment (EzequielMessore) [inline app/build.gradle:68]: On `build.gradle` top level we could use [Extra properties](https://docs.gradle.org/current/userguide/writing_build_scripts.html#sec:extra_properties) to centralize all the versions.

File: app/build.gradle [hunk 2]
```diff
@@ -75,6 +75,12 @@ dependencies {
...
 
     testImplementation 'junit:junit:4.13.2'
+    testImplementation "com.google.truth:truth:1.1.3"
```
Comment (EzequielMessore) [inline app/build.gradle]: What do you need to use these dependencies ?

- com.google.truth:truth
- app.cash.turbine:turbine

File: app/src/test/java/com/example/mybeats/data/repository/MockedProductBodyResponse.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package com.example.mybeats.data.repository
+
+import com.example.mybeats.data.remote.model.ProductBody
+import com.example.mybeats.data.remote.model.ProductsBody
+
+object MockedProductBodyResponse {
```
Comment (EzequielMessore) [inline app/src/test/java/com/example/mybeats/data/repository/MockedProductBodyResponse.kt]: I think you don't need an object, you can just have a function or a variable, but it's up to you.

File: app/src/test/java/com/example/mybeats/view/products/list/ProductsViewModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                assertThat(awaitItem()).isEqualTo(ProductsViewModel.ViewState.Success(mockedProductsList))
+                cancelAndConsumeRemainingEvents()
+            }
+        }
+    }
+
+    @Test
+    fun `getProducts SHOULD emit ViewState Initial, Loading an Error to View WHEN receive ResultRemote Error`(){
+        runBlocking {
+            coEvery { productsRepository.getProducts() } returns flowOf(
```
Comment (EzequielMessore) [inline app/src/test/java/com/example/mybeats/view/products/list/ProductsViewModelTest.kt:96]: If you throws an Exception what’s happen ?

File: app/src/test/java/com/example/mybeats/data/repository/ProductsRepositoryTest.kt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+class ProductsRepositoryTest {
+
+    @get:Rule
+    var rule: TestRule = InstantTaskExecutorRule()
+
+    @MockK
+    private lateinit var productsApi: ProductsApi
+
+    @MockK
+    private lateinit var productsBody: ProductsBody
```
Comment (EzequielMessore) [inline app/src/test/java/com/example/mybeats/data/repository/ProductsRepositoryTest.kt]: The variable was never used, remember to remove import too

```suggestion
```

File: app/build.gradle (hunk 2 above)
Comment (andersonluisp) [reply app/build.gradle]: I used the Truth lib to make the assertThat in the tests. This lib turn the process to verify the asserts more easily.

The Turbine lib it's de only form that I found it to test Hot Flows and in this project I decide to use StateFlow instead LiveData.

File: app/src/test/java/com/example/mybeats/data/repository/MockedProductBodyResponse.kt (hunk 3 above)
Comment (andersonluisp) [reply app/src/test/java/com/example/mybeats/data/repository/MockedProductBodyResponse.kt]: That make senses. I thought about centralizing all possible mocks in one place, but I only used a mocked list.

I'll fix it.

File: app/src/test/java/com/example/mybeats/data/repository/ProductsRepositoryTest.kt (hunk 5 above)
Comment (andersonluisp) [reply app/src/test/java/com/example/mybeats/data/repository/ProductsRepositoryTest.kt]: Sure... I'll remove it.

File: app/src/test/java/com/example/mybeats/view/products/list/ProductsViewModelTest.kt (hunk 4 above)
Comment (andersonluisp) [reply app/src/test/java/com/example/mybeats/view/products/list/ProductsViewModelTest.kt:96]: In this case I failed and not catch exception in the ViewModel.

I will revise it and create the test to this scenario.

File: app/build.gradle (hunk 1 above)
Comment (andersonluisp) [reply app/build.gradle:68]: Oh sure I'll do it.

File: app/build.gradle (hunk 2 above)
Comment (EzequielMessore) [reply app/build.gradle]: The `junit` not was sufficient ?

Nice decision about `Turbine` I had never seen before, I don't got were you using this.

File: app/src/test/java/com/example/mybeats/data/repository/MockedProductBodyResponse.kt (hunk 3 above)
Comment (EzequielMessore) [reply app/src/test/java/com/example/mybeats/data/repository/MockedProductBodyResponse.kt]: Yeah you could centralize this but I think that you can use the another way like:

`object FakeProductsRepository {
    fun getProductsBody() : List<ProductsBody> {
        // return your list.
    }
}`

File: app/build.gradle (hunk 2 above)
Comment (andersonluisp) [reply app/build.gradle]: With junit I could use Assert.assertTrue method for example (Assert.assertTrue(getProductsResult.first() is ResultRemote.Success), but with truth when the test fails I can see this output on terminal that show me what is wrong.

expected instance of: com.example.mybeats.data.remote.responses.ResultRemote$Success
but was instance of : com.example.mybeats.data.remote.responses.ResultRemote$ErrorResponse$Unknown
with value          : Unknown(throwable=java.lang.Throwable)

With junit is it possible too?

File: app/src/test/java/com/example/mybeats/data/repository/MockedProductBodyResponse.kt (hunk 3 above)
Comment (andersonluisp) [reply app/src/test/java/com/example/mybeats/data/repository/MockedProductBodyResponse.kt]: Gotcha. I'll update it.

**Review by EzequielMessore — APPROVED** [review]

**Pull request merged**